### ISSUE TITLE
proposal: Add `GenerateOne` to `encoding/openapi`

### ISSUE THREAD

**Issue by sdboyer** [cue-lang/cue#2076]
**Is your feature request related to a problem? Please describe.**

OpenAPI is the gateway we (Grafana) use for translating CUE to a number of output targets. However, [the current generator](https://pkg.go.dev/cuelang.org/go@v0.5.0-alpha.1/encoding/openapi#Generate) is strongly opinionated about what its inputs look like: it expects a struct, and will generate openapi schema components for each of the definitions it finds within that struct. (And maybe some others that are referenced from within that struct, depending on configuration options.)

**Describe the solution you'd like**

Introduce a `GenerateOne` func to `encoding/openapi`.

In this func, instead of searching the provided `cue.Value` for definitions to generate as elements within `components/schemas`, the generator should just treat the provided `cue.Value` as though it were just such a discovered definition.

This is exactly what i've done in cuetsy, and it's been quite helpful - the main entrypoint, (`GenerateAST`)[https://pkg.go.dev/github.com/grafana/cuetsy#GenerateAST], treats the provided `cue.Value` as a scope in which to search for root objects to translate. [`GenerateSingleAST`](https://pkg.go.dev/github.com/grafana/cuetsy#GenerateSingleAST) treats the provided `cue.Value` as _the one_ object to generate.

Now, cuetsy uses attributes instead of definitions to decide which vertices should be top-level generated types, so it can guarantee a single object output in a way that an openapi generator probably can't. But that's OK anyway, as OpenAPI has no concept of a "root" schema, so a `GenerateOne` has a built-in place to produce...err, more than one. (Maybe the naming here would be better if the current `Generate` were called `GenerateChildren` and this new one were called `Generate`, but that ship has probably sailed at this point)

Without having dug in, i suspect this would be a reasonable signature:

```go
func GenerateOne(sch cue.Value, name string, c *Config)
```

`name` specifies the path of the output schema within `components/schemas`.

**Describe alternatives you've considered**

In the Thema context, we simply don't rely on definitions to indicate "this thing is a schema." So the alternatives are all [hacks](https://github.com/grafana/thema/blob/00ad2949c7bc6b28983afa8d9b4f788332c64469/internal/util/util.go#L70) that map Thema's model onto `encoding/openapi`'s assumptions.

However, i'm now faced with cases where Thema lineages are referencing `import`ed values, which makes that whole print-and-recompile approach untenable. I really need to be able to just hand the `cue.Value` that represents the thing i know to be "a schema" to my generators. I don't really see an alternative to that.

Given an acceptance of the goal, an alternative approach might be another option to `Config` to indicate the "use the root" behavior instead of "search children." However, tbqh i've found those other options fairly confusing, and this is such a binary difference in behavior that i like the cuetsy approach of having two separate functions. Also, `name` would need to become a config property as well, but would only have an effect when using this alternative search pattern.

Really though, either way would be fine 🤷🏻

**Comment by sdboyer** [comment]
To be clear, i'm happy to implement this!

**Comment by rogpeppe** [comment]
To be clear, after discussing online, it seems like what you're after here is a shorthand for calling `openapi.Generate` with a struct containing just one field, `#name` where `name` is the name passed to `GenerateOne` and its value being the `sch` argument.

Something like this, perhaps:
```
func GenerateOne(sch cue.Value, name string, c *openapi.Config) (*ast.File, error) {
	ctx := sch.Context()
	v := ctx.CompileString(fmt.Sprintf(`#%s: _`, name))
	v = v.FillPath(cue.MakePath(cue.Def(name)), sch)
	return openapi.Generate(v, c)
}
```
Would that seem accurate?

**Comment by sdboyer** [comment]
Yep! That would do it.

And yeah, after we discussed in slack, it became clear that the actual problem here was that i'd tried something similar to that implementation, but encountered a panic. I used `FillPath` on a pre-existing `cue.Value` created through `load.Instances` and then `cue.Context.BuildInstance()`, rather than `cue.Context.CompileString`. I'll try to make a full reproducer.

IMO `GenerateOne` is still worth adding because, simplicity of that implementation notwithstanding:

* It keeps folks from falling into the TIMTOWTDI hole that i did, and convincing themselves this isn't well-supported/clearly possible
* It's weird, weird to think about where to document use of `CompileString(fmt.Sprintf("#%s: _", name))` as a standard practice - seems far easier to hide that within a good ol' exported function.
* Would be quite important if the openapi encoder is exposed as a native CUE package (which i hope it is)
